backends/ebpf/ebpfTable.h

```diff
@@ -38,44 +58,84 @@ class EBPFTableBase : public EBPFObject {
     }
 };
 
-class EBPFTable final : public EBPFTableBase {
+class EBPFTable : public EBPFTableBase {
+    const int prefixLenFieldWidth = 32;
+
+    void initKey();
+
+ protected:
+    const cstring prefixFieldName = "prefixlen";
+
+    bool isLPMTable();
+    virtual void validateKeys() const;
+    virtual ActionTranslationVisitor*
+    createActionTranslationVisitor(cstring valueName, const EBPFProgram* program) const {
+        return new ActionTranslationVisitor(valueName, program);
+    }
+
  public:
     const IR::Key*            keyGenerator;
     const IR::ActionList*     actionList;
-    const IR::TableBlock*    table;
+    const IR::TableBlock*     table;
     cstring               defaultActionMapName;
     cstring               actionEnumName;
     std::map<const IR::KeyElement*, cstring> keyFieldNames;
     std::map<const IR::KeyElement*, EBPFType*> keyTypes;
 
     EBPFTable(const EBPFProgram* program, const IR::TableBlock* table, CodeGenInspector* codeGen);
-    void emitTypes(CodeBuilder* builder);
-    void emitInstance(CodeBuilder* builder);
+
+    cstring p4ActionToActionIDName(const IR::P4Action * action) const;
     void emitActionArguments(CodeBuilder* builder, const IR::P4Action* action, cstring name);
-    void emitKeyType(CodeBuilder* builder);
-    void emitValueType(CodeBuilder* builder);
     void emitKey(CodeBuilder* builder, cstring keyName);
-    void emitAction(CodeBuilder* builder, cstring valueName);
-    void emitInitializer(CodeBuilder* builder);
+
+    virtual void emitTypes(CodeBuilder* builder);
+    virtual void emitInstance(CodeBuilder* builder);
+    virtual void emitKeyType(CodeBuilder* builder);
+    virtual void emitValueType(CodeBuilder* builder);
+    virtual void emitValueActionIDNames(CodeBuilder* builder);
+    virtual void emitValueStructStructure(CodeBuilder* builder);
+    virtual void emitAction(CodeBuilder* builder, cstring valueName, cstring actionRunVariable);
+    virtual void emitInitializer(CodeBuilder* builder);
+    virtual void emitLookup(CodeBuilder* builder, cstring key, cstring value) {
+        builder->target->emitTableLookup(builder, dataMapName, key, value);
+        builder->endOfStatement(true);
+    }
+    virtual void emitLookupDefault(CodeBuilder* builder, cstring key, cstring value) {
+        builder->target->emitTableLookup(builder, defaultActionMapName, key, value);
+        builder->endOfStatement(true);
+    }
+    virtual bool isMatchTypeSupported(const IR::Declaration_ID* matchType) {
+        return matchType->name.name == P4::P4CoreLibrary::instance.exactMatch.name ||
+               matchType->name.name == P4::P4CoreLibrary::instance.lpmMatch.name;
+    }
+    // Whether to drop packet if no match entry found.
+    // Some table implementations may want to continue processing.
+    virtual bool dropOnNoMatchingEntryFound() const { return true; }
 };
 
-class EBPFCounterTable final : public EBPFTableBase {
+class EBPFCounterTable : public EBPFTableBase {
```

Comment:
I wonder whether these API changes will impact the p4c-xdp project which also extends the ebpf backend.
Could you actually install that project github.com/vmware/p4c-xdp and run its tests as well?
It should be as easy as creating a symlink in the extensions directory.
In principle I can help with fixes there if you cannot do them.